Do not allocate WTF::Function in hot path

Constellation opened the pull request:
#### 34dcefc88b83e69809fc0115f4173c0289d52196
<pre>
Do not allocate WTF::Function in hot path
<a href="https://bugs.webkit.org/show_bug.cgi?id=273530">https://bugs.webkit.org/show_bug.cgi?id=273530</a>
<a href="https://rdar.apple.com/127332077">rdar://127332077</a>

Reviewed by Keith Miller and Ryosuke Niwa.

We should just use template parameter instead of using Function. It does heap allocation.
If it is hot path and if we do not store the function to somewhere to keep out of this call&apos;s scope, it is wasting.

* Source/WebCore/bindings/js/JSCustomElementInterface.cpp:
(WebCore::JSCustomElementInterface::invokeCallback):
(WebCore::JSCustomElementInterface::invokeConnectedCallback):
(WebCore::JSCustomElementInterface::invokeDisconnectedCallback):
(WebCore::JSCustomElementInterface::invokeFormResetCallback):
* Source/WebCore/bindings/js/JSCustomElementInterface.h:
(WebCore::JSCustomElementInterface::invokeCallback): Deleted.
* Source/WebCore/bindings/js/JSValueInWrappedObject.h:
(WebCore::cachedPropertyValue):

Canonical link: <a href="https://commits.webkit.org/278199@main">https://commits.webkit.org/278199@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c246bdcaaf7f0cab9a2d1563067d6bfe3e7ae2f1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [  ~~🧪 style~~](https://ews-build.webkit.org/#/builders/38/builds/49813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29102 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/2036 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53058 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/59/builds/492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [❌ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52116 "Failed to checkout and rebase branch from PR 27964") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35133 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/59 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | 
| [❌ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51912 "Failed to checkout and rebase branch from PR 27964") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/26650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/2036 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/21765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/24082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/2036 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8185 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/43137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/46058 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/2036 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54639 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/49309 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/24907 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/55 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/48029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/26166 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/2036 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/47057 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/27022 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/56793 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7176 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/25894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11667 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/WebCore/bindings/js/JSCustomElementInterface.cpp:
```diff
@@ -270,7 +270,8 @@ void JSCustomElementInterface::upgradeElement(Element& element)
 
-void JSCustomElementInterface::invokeCallback(Element& element, JSObject* callback, const Function<void(JSGlobalObject*, JSDOMGlobalObject*, MarkedArgumentBuffer&)>& addArguments)
+template<typename Function>
+void JSCustomElementInterface::invokeCallback(Element& element, JSObject* callback, const Function& addArguments)
```
Nit: I think you can just do: 
`void JSCustomElementInterface::invokeCallback(Element& element, JSObject* callback, const auto& addArguments)`

kmiller68 commented on Source/WebCore/bindings/js/JSCustomElementInterface.h:
```diff
@@ -124,7 +124,8 @@ class JSCustomElementInterface : public RefCounted<JSCustomElementInterface>, pu
 
-    void invokeCallback(Element&, JSC::JSObject* callback, const Function<void(JSC::JSGlobalObject*, JSDOMGlobalObject*, JSC::MarkedArgumentBuffer&)>& addArguments = [](JSC::JSGlobalObject*, JSDOMGlobalObject*, JSC::MarkedArgumentBuffer&) { });
+    template<typename Function>
+    void invokeCallback(Element&, JSC::JSObject* callback, const Function& addArguments);
```
Ditto.

kmiller68 commented on Source/WebCore/bindings/js/JSValueInWrappedObject.h:
```diff
@@ -56,7 +56,8 @@ class JSValueInWrappedObject {
 
-JSC::JSValue cachedPropertyValue(JSC::ThrowScope&, JSC::JSGlobalObject&, const JSDOMObject& owner, JSValueInWrappedObject& cacheSlot, const Function<JSC::JSValue(JSC::ThrowScope&)>&);
+template<typename Function>
+JSC::JSValue cachedPropertyValue(JSC::ThrowScope&, JSC::JSGlobalObject&, const JSDOMObject& owner, JSValueInWrappedObject& cacheSlot, const Function&);
```
Ditto.

kmiller68 commented on Source/WebCore/bindings/js/JSValueInWrappedObject.h:
```diff
@@ -109,7 +110,8 @@ inline void JSValueInWrappedObject::clear()
 
-inline JSC::JSValue cachedPropertyValue(JSC::ThrowScope& throwScope, JSC::JSGlobalObject& lexicalGlobalObject, const JSDOMObject& owner, JSValueInWrappedObject& cachedValue, const Function<JSC::JSValue(JSC::ThrowScope&)>& function)
+template<typename Function>
+inline JSC::JSValue cachedPropertyValue(JSC::ThrowScope& throwScope, JSC::JSGlobalObject& lexicalGlobalObject, const JSDOMObject& owner, JSValueInWrappedObject& cachedValue, const Function& function)
```
Ditto.

kmiller68 reviewed (approved):
r=me with nits.

Constellation committed:
Do not allocate WTF::Function in hot path
https://bugs.webkit.org/show_bug.cgi?id=273530
rdar://127332077

Reviewed by Keith Miller and Ryosuke Niwa.

We should just use template parameter instead of using Function. It does heap allocation.
If it is hot path and if we do not store the function to somewhere to keep out of this call's scope, it is wasting.

* Source/WebCore/bindings/js/JSCustomElementInterface.cpp:
(WebCore::JSCustomElementInterface::invokeCallback):
(WebCore::JSCustomElementInterface::invokeConnectedCallback):
(WebCore::JSCustomElementInterface::invokeDisconnectedCallback):
(WebCore::JSCustomElementInterface::invokeFormResetCallback):
* Source/WebCore/bindings/js/JSCustomElementInterface.h:
(WebCore::JSCustomElementInterface::invokeCallback): Deleted.
* Source/WebCore/bindings/js/JSValueInWrappedObject.h:
(WebCore::cachedPropertyValue):

Canonical link: https://commits.webkit.org/278199@main
